hardware_interface/src/utils/component_parser.cpp

```diff
@@ -0,0 +1,200 @@
+// Copyright 2020 ros2_control Development Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <tinyxml2.h>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+
+#include "hardware_interface/component_info.hpp"
+#include "hardware_interface/utils/component_parser.hpp"
+
+namespace
+{
+constexpr const auto kRobotTag = "robot";
+constexpr const auto kROS2ControlTag = "ros2_control";
+constexpr const auto kHardwareTag = "hardware";
+constexpr const auto kClassTypeTag = "classType";
+constexpr const auto kParamTag = "param";
+constexpr const auto kJointTag = "joint";
+constexpr const auto kInterfaceNameTag = "interfaceName";
+
+// For complete reference of syntax - not used in parser
+//  constexpr const auto kJointTag = "joint";
+//  constexpr const auto kSensorTag = "sensor";
+}  // namespace
+
+namespace hardware_interface
+{
+namespace utils
+{
+
+SystemInfo parse_system_from_urdf(const std::string & urdf)
+{
+  // Check if everything OK with URDF string
+  if (urdf.empty()) {
+    throw std::runtime_error("empty URDF passed to robot");
+  }
+  tinyxml2::XMLDocument doc;
+  if (!doc.Parse(urdf.c_str()) && doc.Error()) {
+    throw std::runtime_error("invalid URDF passed in to robot parser");
+  }
+
+  // Find robot tag
+  const tinyxml2::XMLElement * robot_it = doc.RootElement();
+
+  if (std::string(kRobotTag).compare(robot_it->Name())) {
+    throw std::runtime_error("the robot tag is not root element in URDF");
+  }
+
+  SystemInfo system;
+  const tinyxml2::XMLAttribute * attr;
+
+  attr = robot_it->FindAttribute("name");
+  if (!attr) {
+    throw std::runtime_error("no robot name attribute set");
+  }
+  system.name = robot_it->Attribute("name");
+
+  const tinyxml2::XMLElement * ros2_control_it = robot_it->FirstChildElement(kROS2ControlTag);
+  if (!ros2_control_it) {
+    throw std::runtime_error("no " + std::string(kROS2ControlTag) + " tag");
+  }
+  attr = ros2_control_it->FindAttribute("name");
```

Comment:
I think this section could be made more readable if you had a helper function that's something along these lines:

```
string get_attr_value(const tinyxml2::XMLElement * some_it, string attribute_name)
{
const tinyxml2::XMLAttribute * attr;
attr = some_it->FindAttribute(attribute_name);
  if (!attr) {
    throw std::runtime_error("no attribute name in " + attribute_name + " tag");
  }
return some_it->Attribute(attribute_name);
}
```
then you can simplify this parsing bit to:
```
// shorter version of getting the attribute value
  system.type = get_attr_value(ros2_control_it, "type");
```
I'm not sure this applies to all cases but it seems like a fair bit could use it